src/nanovg_pick.c

```diff
@@ -0,0 +1,1742 @@
+static void nvg__splitBezier(const float* points, float t, float* pointsA, float* pointsB);
+static void nvg__smallsort(float* values, int n);
+static void nvg__pickSceneInsert(NVGpickScene* ps, NVGpickPath* pp);
+static NVGpickScene* nvg__pickSceneGet(NVGcontext* ctx);
+
+//typedef struct NVGpickScene NVGpickScene;
+
+//
+// Bounds Utilities
+//
+
+static void nvg__initBounds(float bounds[4])
+{
+	bounds[0] = bounds[1] = 1e6f;
+	bounds[2] = bounds[3] = -1e6f;
+}
+
+static void nvg__expandBounds(float bounds[4], const float *points, int npoints)
+{
+	npoints *= 2;
+	int i;
+	for (i = 0; i < npoints; i += 2)
+	{
+		bounds[0] = nvg__minf(bounds[0], points[i]);
+		bounds[1] = nvg__minf(bounds[1], points[i+1]);
+		bounds[2] = nvg__maxf(bounds[2], points[i]);
+		bounds[3] = nvg__maxf(bounds[3], points[i+1]);
+	}
+}
+
+static void nvg__unionBounds(float bounds[4], const float boundsB[4])
+{
+	bounds[0] = nvg__minf(bounds[0], boundsB[0]);
+	bounds[1] = nvg__minf(bounds[1], boundsB[1]);
+	bounds[2] = nvg__maxf(bounds[2], boundsB[2]);
+	bounds[3] = nvg__maxf(bounds[3], boundsB[3]);		
+} 
+
+static void nvg__intersectBounds(float bounds[4], const float boundsB[4])
+{
+	bounds[0] = nvg__maxf(boundsB[0], bounds[0]);
+	bounds[1] = nvg__maxf(boundsB[1], bounds[1]);
+	bounds[2] = nvg__minf(boundsB[2], bounds[2]);
+	bounds[3] = nvg__minf(boundsB[3], bounds[3]);
+	
+	bounds[2] = nvg__maxf(bounds[0], bounds[2]);
+	bounds[3] = nvg__maxf(bounds[1], bounds[3]);
+}
+
+//
+// Building paths & sub paths
+//
+
+static int nvg__pickSubPathAddPoints(NVGpickScene* ps, NVGpickSubPath* psp, const float *xy, int n)
+{
+	if (ps->npoints + n > ps->cpoints)
+	{
+		int cpoints = ps->npoints + n + (ps->cpoints << 1);
+		float* points = realloc(ps->points, sizeof(float) * 2 * cpoints);
+		if (points == NULL) return -1;
+		ps->points = points;
+		ps->cpoints = cpoints;
+	}
+	int i = ps->npoints;
+	memcpy(&ps->points[i * 2], xy, sizeof(float) * 2 * n);
+	ps->npoints += n;
+	return i;
+}
+
+static void nvg__pickSubPathAddSegment(NVGpickScene* ps, NVGpickSubPath* psp, int firstPoint, int type, short flags)
+{
+	if (ps->nsegments == ps->csegments)
+	{
+		int csegments = 1 + ps->csegments + (ps->csegments << 1);
+		NVGsegment* segments = (NVGsegment*)realloc(ps->segments, sizeof(NVGsegment) * csegments);
+		if (segments == NULL) return;
+		ps->segments = segments;
+		ps->csegments = csegments;
+	}
+	
+	if (psp->firstSegment == -1)	
+		psp->firstSegment = ps->nsegments;
+		
+	NVGsegment* seg = &ps->segments[ps->nsegments];
+	++ps->nsegments;
+	seg->firstPoint = firstPoint;
+	seg->type = (short)type;
+	seg->flags = flags;
+	++psp->nsegments;			
+
+	nvg__segmentDir(ps, psp, seg,  0, seg->startDir);
+	nvg__segmentDir(ps, psp, seg,  1, seg->endDir);	
+}
+
+static void nvg__segmentDir(NVGpickScene* ps, NVGpickSubPath* psp, NVGsegment* seg, float t, float d[2])
+{
+	#define PosX(i) points[i*2]
```

Comment:
Prefer temp variables instead of macros like these, i.e.:
```
float x0 = points[0*2+0], x1 = points[1*2+0], x2 = points[2*2+0];
float y0 = points[0*2+1], y1 = points[1*2+1], y2 = points[2*2+1];
```